### ISSUE TITLE
Refactor to TypeScript, Documentation, NPM and More

### ISSUE THREAD

**Issue by laobubu** [laobubu/HyperMD#21]
- [x] 🏠 HyperMD Core 
- [x] ⚡️  HyperMD mode
- [x] 🔌 Addons 
   - [x] InsertFile
   - [x] Click
   - [x] CursorDebounce
   - [x] FoldMath
   - [x] Fold
   - [x] HideToken
   - [x] Hover
   - [x] ModeLoader
   - [x] Paste
   - [x] ReadLink
   - [x] TableAlign
- [x] ⚙️ Bundler config (with rollup.js) 
- [x] 📑  Interfaces Declaration
- [x] 🙃 Deal with the poorly-typed [@types/codemirror](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/codemirror)
- [ ] 📘  Documentation
    - [ ] CONTRIBUTING.md
    - [x] Doc generator with TypeScript
- [x] 📦  Pack-up script
- [x] 🚢  First Release! Publish to npm

**Comment by laobubu** [comment]
Code goes to [refactoring-with-ts branch](https://github.com/laobubu/HyperMD/tree/refactoring-with-ts)

Rollup is good but there is still something wrong in my configuration...

> Emmm... **what the hell is `'../../hypermd/addon/'`  ?!**
> https://github.com/laobubu/HyperMD/blob/f108ab1e037b8d0df81e60870278bb3b5ac5d449/dist/addon/insert-file.js#L2-L4

@nokola

**Comment by nokola** [comment]
re: '../../hypermd/addon/' -> looks like you fixed it, probably some issue with import or base folder.

"Deal with the poorly-typed @types/codemirror": I noticed for some function calls I had to replace them to use Doc instead of Editor. e.g. cm.getCursor() change to cm.getDoc().getCursor(). I think this is intentional because CodeMirror.Doc is the better location to deal with cursor/etc. The old functions are left working only for legacy purposes.

**Comment by laobubu** [comment]
A workaround is that let CodrMirror.Editor extends CodeMirror.Doc , which is now written in "src/core/type.d.ts". I'm not sure whether multi-doc feature is important. If it is, lots of add-ons will need to rewrite. So let's just put this aside for a whil.

**Comment by nokola** [comment]
Just found something very interesting! See parcel js - a module bundler with what they say very good typescript support https://parceljs.org/transforms.html

I haven't read much about it yet.

**Comment by laobubu** [comment]
# (part of) BREAKING CHANGES

## Powerful *Fold* System!

`Fold` addon is now much more efficient and extensible. 

A developer may implement a `FolderFunc` to fold everything he want. A example can be found in [FoldMath, only 60 lines(with comments)](https://github.com/laobubu/HyperMD/blob/25b44316a537fdc115515c99a9527114889d7eff/src/addon/fold-math.ts#L15-L75).

This makes it very easy to support [Markdown Diagrams](https://github.com/shd101wyy/markdown-preview-enhanced/blob/master/docs/diagrams.md) and render them with 3rd libraries like [flowchart.js](http://flowchart.js.org/), [PlantUML](http://plantuml.com/) etc.

https://github.com/laobubu/HyperMD/blob/25b44316a537fdc115515c99a9527114889d7eff/src/addon/fold.ts#L17-L29

## Click Handers

`hmdClick` option can be a ClickHandler function, which may supress HyperMD default behaviors by returning `false`:

https://github.com/laobubu/HyperMD/blob/25b44316a537fdc115515c99a9527114889d7eff/src/addon/click.ts#L17-L38

## Addon Skeleton

Boilerplate to start a new add-on quickly. See [skeleton.ts](https://github.com/laobubu/HyperMD/blob/25b44316a537fdc115515c99a9527114889d7eff/src/addon/skeleton.ts)

Once a new add-on is ready-to-test:

1. Add to `components` in *rollup.config.js*
2. Edit *demo/index.js* and load it
3. Edit *src/core/quick.ts* and provide a suggested option value
4. `yarn dev`

-------------------

@nasyxx interested in composing **CONTRIBUTING.md**? I do feel frustrated with my poor English now.

**Comment by cgestes** [comment]
Happy to unroll my custom change to make it work in webpack and try this changeset :)

Will keep you posted.

**Comment by cgestes** [comment]
using your branch at commit `955fffe`

```bash
$ yarn build                                                                                                                                                                           19:25
yarn run v1.6.0
$ npm run build_js && npm run build_css

> hypermd@0.3.0 build_js /home/cedric/src/HyperMD
> rollup -c


./src/core.ts → ./core.js...
[!] (rpt2 plugin) Error: /home/cedric/src/HyperMD/src/core.ts(5,24): semantic error TS2307 Cannot find module './core/Addon'.
src/core.ts
Error: /home/cedric/src/HyperMD/src/core.ts(5,24): semantic error TS2307 Cannot find module './core/Addon'.
    at error (/home/cedric/src/HyperMD/node_modules/rollup/dist/rollup.js:199:15)
    at Object.error (/home/cedric/src/HyperMD/node_modules/rollup/dist/rollup.js:17311:21)
    at RollupContext.error (/home/cedric/src/HyperMD/node_modules/rollup-plugin-typescript2/dist/rollup-plugin-typescript2.cjs.js:17215:30)
    at /home/cedric/src/HyperMD/node_modules/rollup-plugin-typescript2/dist/rollup-plugin-typescript2.cjs.js:19860:23
    at arrayEach (/home/cedric/src/HyperMD/node_modules/rollup-plugin-typescript2/dist/rollup-plugin-typescript2.cjs.js:546:11)
    at forEach (/home/cedric/src/HyperMD/node_modules/rollup-plugin-typescript2/dist/rollup-plugin-typescript2.cjs.js:9372:14)
    at printDiagnostics (/home/cedric/src/HyperMD/node_modules/rollup-plugin-typescript2/dist/rollup-plugin-typescript2.cjs.js:19833:5)
    at Object.transform (/home/cedric/src/HyperMD/node_modules/rollup-plugin-typescript2/dist/rollup-plugin-typescript2.cjs.js:20104:17)
    at /home/cedric/src/HyperMD/node_modules/rollup/dist/rollup.js:17316:48
    at process._tickCallback (internal/process/next_tick.js:68:7)
```

**Comment by laobubu** [comment]
@cgestes 
I use Windows and filenames are case-insensitive. Thanks for reporting this bug.
The package on NPM is now tested (with `parcel-bundler`) and works well. See <http://laobubu.net/HyperMD/#./docs/index.md>

**Comment by cgestes** [comment]
Ok thank you. I'll see it seems there is some migration required. I'll keep you posted

Thanks for the great addons.

**Comment by nokola** [comment]
Will you merge the typescript branch before the documentation? I think the documentation can be a separate task - for me typescript support is an awesome thing of it's own :)
